### DISCUSSION TITLE
[Bug] error MSB6006: "java" exited with code 2 - Cannot deploy to device

### DISCUSSION THREAD

**Discussion by theolliebbb** [xamarin/Xamarin.Forms#15293] [Q&A]
### Description
(copied from stackoverflow) https://stackoverflow.com/questions/71751365/error-msb6006-java-exited-with-code-2-on-xamarin-android?noredirect=1#comment126818451_71751365

I know this question has been asked several times- My Xamarin Android package builds just fine, and gives me this error- I can't run ANY project (including fresh new projects) on any emulator or my own physical Android- Runs totally fine on IOS. I'm on Mac

I've tried everything I could find online

- Reinstalling VS 2019

- Using VS 2022

- Reinstalling Xamarin and all emulators as well as Java SDKs and Android SDKs

- Enabling/disabling multidex or Appkey signing with debug.keystore and other custom made keys ( this gives error MSB6006: "java" exited with code 1.) on both debug and release

Sometimes with different parameters, I get "error ADB0010 The package was not properly signed (NO_CERTIFICATES)."

Also not able to find a solution for that

This problem came about after updating an app on google play console, but I don't see why this would have anything to do with it

If anybody could help, it'd be greatly appreciated- I'm completely unable to test my xamarin projects on Android at all and it's driving me crazy.

Edit:
The code that is supposedly the culprit is:
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.targets(2309,2)

Which is 

    <AndroidApkSigner Condition=" '$(AndroidUseApkSigner)' == 'true' "
		ApkSignerJar="$(ApkSignerJar)"
		ApkToSign="%(ApkAbiFilesAligned.Identity)"
		KeyStore="$(_ApkKeyStore)"
		KeyAlias="$(_ApkKeyAlias)"
		KeyPass="$(_ApkKeyPass)"
		StorePass="$(_ApkStorePass)"
		PlatformKey="$(AndroidSigningPlatformKey)"
		PlatformCert="$(AndroidSigningPlatformCert)"
		ToolPath="$(JavaToolPath)"
		ToolExe="$(JavaToolExe)"
		ManifestFile="$(IntermediateOutputPath)android\AndroidManifest.xml"
		AdditionalArguments="$(AndroidApkSignerAdditionalArguments)"
	/>

Edit 2:

When ASB00100 shows up here's the error log:
``
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010: Mono.AndroidTools.InstallFailedException: The package was not properly signed (NO_CERTIFICATES).
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010:   at Mono.AndroidTools.Internal.AdbOutputParsing.CheckInstallSuccess (System.String output, System.String packageName) [0x000de] in /Users/builder/azdo/_work/1/s/xamarin-android/external/monodroid/tools/msbuild/external/androidtools/Mono.AndroidTools/Internal/AdbOutputParsing.cs:338 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010:   at Mono.AndroidTools.AndroidDevice+<>c__DisplayClass100_0.<InstallPackage>b__0 (System.Threading.Tasks.Task`1[TResult] t) [0x00016] in /Users/builder/azdo/_work/1/s/xamarin-android/external/monodroid/tools/msbuild/external/androidtools/Mono.AndroidTools/AndroidDevice.cs:803 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010:   at System.Threading.Tasks.ContinuationTaskFromResultTask`1[TAntecedentResult].InnerInvoke () [0x00024] in <e068e2227ab74c1bb3d724ebaab0e3ff>:0 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010:   at System.Threading.Tasks.Task.Execute () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:2319 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010: --- End of stack trace from previous location where exception was thrown ---
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010: 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010:   at AndroidDeviceExtensions.PushAndInstallPackageAsync (Mono.AndroidTools.AndroidDevice device, PushAndInstallCommand command, System.Threading.CancellationToken token) [0x001cf] in /Users/builder/azdo/_work/1/s/xamarin-android/external/monodroid/tools/msbuild/external/androidtools/Xamarin.AndroidTools/Devices/AndroidDeviceExtensions.cs:206 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010:   at AndroidDeviceExtensions.PushAndInstallPackageAsync (Mono.AndroidTools.AndroidDevice device, PushAndInstallCommand command, System.Threading.CancellationToken token) [0x0042f] in /Users/builder/azdo/_work/1/s/xamarin-android/external/monodroid/tools/msbuild/external/androidtools/Xamarin.AndroidTools/Devices/AndroidDeviceExtensions.cs:223 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010:   at Xamarin.Android.Tasks.FastDeploy.InstallPackage (System.Boolean installed) [0x001c0] in /Users/builder/azdo/_work/1/s/xamarin-android/external/monodroid/tools/msbuild/Tasks/FastDeploy.cs:338 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010:   at Xamarin.Android.Tasks.FastDeploy.InstallPackage (System.Boolean installed) [0x0032d] in /Users/builder/azdo/_work/1/s/xamarin-android/external/monodroid/tools/msbuild/Tasks/FastDeploy.cs:355 
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.Debugging.targets(604,5): error ADB0010:   at Xamarin.Android.Tasks.FastDeploy.RunTaskAsync () [0x002fb] in /Users/builder/azdo/_work/1/s/xamarin-android/external/monodroid/tools/msbuild/Tasks/FastDeploy.cs:212 
``

culprit (Common.Debugging.Targets):

        <FastDeploy
      Condition=" '$(AndroidPackageFormat)' != 'aab' "
      AdbTarget="$(AdbTarget)"
      AllowDeltaInstall="$(_AndroidAllowDeltaInstall)"
      DiagnosticLogging="$(_FastDeploymentDiagnosticLogging)"
      FastDevFiles="@(_FilteredFastDevFiles);@(_ResolvedSymbols);@(_ResolvedConfigFiles)"
      FastDevToolPath="$(MSBuildThisFileDirectory)\lib"
      EmbedAssembliesIntoApk="$(EmbedAssembliesIntoApk)"
      PrimaryCpuAbi="$(_PrimaryCpuAbi)"
      PreserveUserData="$(AndroidPreserveUserData)"
      PackageName="$(_AndroidPackage)"
      PackageFile="$(_ApkToInstall)"
      ReInstall="$(_ReInstall)"
      ToolVersion="$(AndroidFastDeploymentToolVersion)"
      UploadFlagFile="$(_UploadFlag)"
      UsingAndroidNETSdk="$(UsingAndroidNETSdk)"
      UserID="$(AndroidDeviceUserId)"
    />

### Steps to Reproduce

1. Build app to Android Emulator or Physical Android

### Expected Behavior
- It should deploy
### Basic Information

- Platform Target Frameworks: 
  - Android: Any
- Android Support Library / AndroidX Version: 
- NuGet Packages: Any
- Affected Devices: All Androids

### Environment

<details><summary>Show/Hide Visual Studio info</summary>

```
=== Visual Studio Community 2019 for Mac ===

Version 8.10.21 (build 4)
Installation UUID: 01e337df-5109-46ad-8de3-d919620ed742
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 6.18.0.23 (d16-6 / 088c73638)

	Package version: 612000162

=== Mono Framework MDK ===

Runtime:
	Mono 6.12.0.162 (2020-02/2ca650f1f62) (64-bit)
	Package version: 612000162

=== Roslyn (Language Service) ===

3.10.0-4.21269.26+029847714208ebe49668667c60ea5b0a294e0fcb

=== NuGet ===

Version: 5.9.0.7134

=== .NET SDK (x64) ===

SDK: /usr/local/share/dotnet/sdk/6.0.201/Sdks
SDK Versions:
	6.0.201
	5.0.406
	3.1.417
MSBuild SDKs: /Applications/Visual Studio.app/Contents/Resources/lib/monodevelop/bin/MSBuild/Current/bin/Sdks

=== .NET Core Runtime ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	6.0.3
	5.0.15
	3.1.23

=== .NET Core 3.1 SDK ===

SDK: 3.1.417

=== .NET 5.0 SDK ===

SDK: 5.0.406

=== Xamarin.Profiler ===

Version: 1.7.0.0
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Apple Developer Tools ===

Xcode 12.5.1 (18212)
Build 12E507

=== Xamarin.Mac ===

Xamarin.Mac not installed. Can't find /Library/Frameworks/Xamarin.Mac.framework/Versions/Current/Version.

=== Xamarin Designer ===

Version: 16.11.0.60
Hash: 56f9b80b0
Branch: remotes/origin/d16-11
Build date: 2021-12-15 01:44:16 UTC

=== Xamarin.iOS ===

Version: 15.8.0.0 (Visual Studio Community)
Hash: f10d9e023
Branch: xcode13.3
Build date: 2022-03-15 11:47:35-0400

=== Xamarin.Android ===

Version: 12.0.0.3 (Visual Studio Community)
Commit: xamarin-android/d16-11/f0e3c2d
Android SDK: /Users/oll/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		None installed

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 31.0.3
SDK Build Tools Version: 30.0.3

Build Information: 
Mono: c633fe9
Java.Interop: xamarin/java.interop/d16-11@476bb5b
ProGuard: Guardsquare/proguard/v7.0.1@912d149
SQLite: xamarin/sqlite/3.35.4@85460d3
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-11@87af37b

=== Eclipse Temurin JDK ===

Java SDK: Not Found

=== Android SDK Manager ===

Version: 16.10.0.13
Hash: 1b81df5
Branch: remotes/origin/d16-10
Build date: 2021-11-12 00:17:32 UTC

=== Android Device Manager ===

Version: 16.10.0.15
Hash: 89dcc0b
Branch: remotes/origin/d16-10
Build date: 2021-11-12 00:17:52 UTC

=== Build Information ===

Release ID: 810210004
Git revision: eb0b2f7259d35b7d767c79d91d356881227e0985
Build date: 2022-03-17 17:18:28-04
Build branch: release-8.10

=== Operating System ===

Mac OS X 11.6.5
Darwin 20.6.0 Darwin Kernel Version 20.6.0
    Tue Feb 22 21:10:41 PST 2022
    root:xnu-7195.141.26~1/RELEASE_X86_64 x86_64

```

</details>

### Build Logs

Build runs just fine!
here's the deploy to device log with the NO_CERTIFICATES error:
https://gist.github.com/theolliebbb/6eebb9f264b5616828c00384c4275d84
with error code 2:
https://gist.github.com/theolliebbb/ada1ab7f7bb6ca911471843f96e5ba4f

### Reproduction Link

Seems to be specific to my Mac- Any Xamarin project going to android gives this error

**Comment by jfversluis** [comment]
DId you also try things like these? https://stackoverflow.com/questions/63530855/xamarin-app-cant-deploy-to-android-11-emulator

I think ultimately this is a problem in your setup and more a Xamarin.Android thing

**Reply by theolliebbb** [reply]
Thanks for responding-
Yes, I tried that one as well, I'm afraid

**Reply by jfversluis** [reply]
This is one of those things that is very hard to debug if there is no reproduction and we can't just sit next to each other to figure it out I guess 😅

**Reply by theolliebbb** [reply]
I understand-
Luckily on a whim, I checked to see if I had an older debug.keystore to replace the current one with. The current one was a size of 0kb and something was wrong with it.
Doing that fixed it and I can test my Android apps again, so it's all fixed! 
Thanks for trying to help- Big fan of your youtube channel, btw.

**Reply by jfversluis** [reply]
Oh that is wonderful! Glad to hear you still got it resolved and thank you for the kind words!
